### ISSUE TITLE
eval-harness: Cohen's-κ judge calibration before #592 close

### ISSUE THREAD

**Issue by robotrocketscience** [robotrocketscience/aelfrice#687]
## Context

#592 hot-start ≥80% AC verdict ([comment](https://github.com/robotrocketscience/aelfrice/issues/592#issuecomment-4425568029)) was produced from a single Sonnet 4.6 judge call over 18 deduplicated `(expected, actual)` pairs. `docs/BENCHMARKS.md` is explicit that the eval-judge calibration target is **Cohen's-κ disagreement against a zero-LLM baseline** — a single judge run does not establish that.

Operator disposition (2026-05-11): accept the #592 verdict numbers as-is, but gate the issue's close on κ-validation work tracked here.

## Scope

- **Multi-judge run.** Re-run the host-agent eval-replay against `benchmarks/results/592_run/` (or fresh sweep) at least 3× across independent judge invocations. Capture per-call verdicts and check inter-rater agreement.
- **Zero-LLM baseline.** Establish the deterministic baseline the κ measure compares against. If `score_substring_exact_match` (cf. `tests/test_bench_qa_scoring.py`) or another existing path applies, wire it in; otherwise spec the baseline.
- **κ verdict.** Compute Cohen's κ on the matched/unmatched binary verdicts (judge vs. baseline; and pairwise across judge runs for inter-judge agreement). Threshold for "calibrated" goes here — needs operator ratification of the bar before the run.
- **Re-confirm hot-start AC.** Once κ-validated judge produces a stable verdict, confirm hot-start fidelity ≥80% holds across the multi-run sample (not just one run).

## Acceptance

- [ ] Multi-judge run protocol documented in `docs/BENCHMARKS.md` § Eval-judge calibration.
- [ ] Zero-LLM baseline defined and runnable.
- [ ] Cohen's-κ ≥ <threshold-TBD> reported in `benchmarks/results/<run-id>/judge_kappa.json`.
- [ ] Hot-start fidelity ≥80% holds across N≥3 independent judge runs at threshold ≤0.6.
- [ ] #592 unblocked to close on completion.

## Related

- #592 — eval-harness commit-3, AC verdict pending κ-validation.
- #587 — post-compact hot-start; the AC bullet this gates.
- `benchmarks/results/592_run/` — single-run artifacts; re-use as the corpus seed if useful.
- `docs/BENCHMARKS.md` § Pass 1 / Pass 2 — contamination protocol the κ work must respect.

## Out of scope

- Replacing the rebuilder retrieval (still locked deterministic per #605).
- Importing a vendor SDK into aelfrice or its bench code (#607 lock).
- Fixture-convention rework (eval_turns indexing user vs assistant turns) — that's a separate question; if the operator picks option (b) on the #592 thread, this issue may be moot.

**Comment by robotrocketscience** [comment]
## #687 bench plan — proposed protocol (survey, no code)

Filing this as a concrete bench plan ahead of the run. Three operator decisions still need to be ratified before the harness fires (listed at the bottom).

### What gets measured

The judge-augmented hot-start fidelity for the #592 acceptance criterion (`>= 80% on synthetic post-compact scenarios`), validated by Cohen's-kappa across:

1. **Inter-judge agreement.** Same prompt, same `judge_requests.jsonl`, N independent dispatches. Tests judge stability — if kappa is low, the headline number is variance, not signal.
2. **Judge-vs-baseline agreement.** Judge verdict against a deterministic zero-LLM baseline. Tests judge calibration — if kappa is high vs baseline, the judge isn't adding signal beyond the deterministic path; if it's near zero, the judge and baseline disagree systematically (which is the expected case here, since the judge runs *only* on rows the substring path already scored as no-match).

### Corpus

Reuse `benchmarks/results/592_run/` artifacts (per the issue body's "or fresh sweep"). The reusable rows are the judge-eligible subset — replay rows tagged `reason="needs_llm_judge"` from the deterministic pass. Per locked memory the prior 592 bench produced 18 deduplicated `(expected, actual)` pairs from a single judge call; this protocol replays the same subset N times.

If sample size is too small for kappa to stabilise (n=18 is borderline), spec a fresh sweep with `--max-judge-calls` raised to expand the eligible set first.

### Protocol — per run

```
# Setup (once)
RUN_DIR=benchmarks/results/687_kappa_$(date -u +%Y%m%d_%H%M%S)
mkdir -p $RUN_DIR
cp benchmarks/results/592_run/judge_requests.jsonl $RUN_DIR/

# Dispatch N times (operator-driven, host CLI off-band — no SDK in aelfrice per #607 lock)
for i in $(seq 1 $N); do
  # operator dispatches anchor-tier judge against $RUN_DIR/judge_requests.jsonl
  # writes $RUN_DIR/judge_responses_run${i}.jsonl
done

# Compute kappa + per-run fidelity (new module, see "code that needs to ship")
uv run python -m benchmarks.kappa_validation \
    --run-dir $RUN_DIR \
    --baseline f1 --baseline-threshold 0.5 \
    --report $RUN_DIR/judge_kappa.json
```

### Zero-LLM baseline definition

The issue body suggests `score_substring_exact_match` (`benchmarks/qa_scoring.py:53`). On the judge-eligible subset that scorer is **degenerate** — every eligible row was tagged `needs_llm_judge` precisely *because* the substring path returned 0. Using it as the kappa baseline means baseline = always-zero, and Cohen's-kappa against an always-zero rater is undefined / trivially zero. That doesn't produce a meaningful calibration number.

Two options that do work:

- **(a) F1 ≥ tau on the judge-eligible subset.** `score_f1` from the same module is a non-trivial deterministic signal even on rows where the substring path failed (token overlap captures partial answers). Pick tau and binarise. Recommend tau=0.5 for "load-bearing claim is present"; the issue's prompt wording (matched = "candidate conveys the same factual content") aligns roughly with that bar.
- **(b) Run kappa on the *full* eval corpus, not the judge-eligible subset.** Baseline = `score_substring_exact_match` per row (1.0 on substring hits, 0.0 elsewhere); judge-augmented = `apply_judge_verdicts` output. Gives a real binary distribution on both sides. Trade-off: the kappa gets diluted because most full-corpus rows agree trivially (both raters say match on substring hits). Read carefully.

Recommend **(a)** as the primary, **(b)** as a secondary report. Both numbers go into `judge_kappa.json`.

### Math

```
# Per pair (run_i, run_j) for inter-judge:
kappa_ij = cohens_kappa(matched_run_i, matched_run_j)
# Aggregate inter-judge: mean over all C(N,2) pairs.

# Judge-vs-baseline (per run):
kappa_baseline_i = cohens_kappa(matched_run_i, baseline_matched)
# Aggregate: mean over runs.
```

Standard `sklearn.metrics.cohen_kappa_score` is fine; aelfrice doesn't currently depend on sklearn so the kappa computation should be a 20-line stdlib helper in `benchmarks/kappa_validation.py` (formula below).

```
kappa = (p_o - p_e) / (1 - p_e)
where p_o = sum(rater1[i] == rater2[i]) / n
      p_e = sum_k (P(rater1=k) * P(rater2=k))   # k in {0,1}
```

Confidence interval: bootstrap 2000 resamples, report 95% CI. Useful when n is small.

### Acceptance computation

Per the AC bullets:

- **#687 AC-3:** kappa-inter-judge ≥ tau_inter AND kappa-vs-baseline ≥ tau_baseline (operator picks both thresholds, see decisions below).
- **#687 AC-4:** Hot-start fidelity (full eval corpus, judge-augmented) ≥ 80% holds across all N runs at trigger_threshold ≤ 0.6. Per-run fidelity reported alongside median + min.

If both AC-3 and AC-4 pass, #592 unblocks for close.

### Code that needs to ship before the run

1. `benchmarks/kappa_validation.py` — reads `judge_responses_run*.jsonl` from a run dir, joins by `turn_idx`, computes pairwise inter-judge kappa + judge-vs-baseline kappa + bootstrap CI, writes `judge_kappa.json`. Pure stdlib, no sklearn.
2. `tests/test_kappa_validation.py` — known-input fixtures for the kappa formula, bootstrap stability, missing-response handling (run dropped if any response file is incomplete), degenerate baseline detection (refuse to compute against an always-zero rater, surface a clear error).
3. `docs/BENCHMARKS.md § Eval-judge calibration` — append the protocol (this comment, sanitised).

Estimated scope: one PR, ~150 LOC, `rook` tier, can be authored against current main without waiting on any other v3 work. Stacks cleanly under #687 close.

### Operator decisions needed before the run fires

1. **N (number of independent judge runs).** Issue body says "≥3"; recommend **5** for stable kappa CI on n=18 sample. Lower bound 3, upper bound 7 (cost vs. precision).
2. **kappa-inter-judge threshold.** Standard reading: ≥0.61 substantial, ≥0.81 almost-perfect (Landis & Koch 1977). Recommend **≥0.80** since this is judge-stability and we need the headline number to be repeatable.
3. **kappa-vs-baseline threshold.** Lower bar — judge SHOULD diverge from baseline on this subset. Recommend **≥0.40** (moderate); if it drops below, the judge is essentially flipping coins relative to the deterministic signal, which would be a real concern.
4. **F1 threshold tau for baseline.** Recommend **0.5**.
5. **Host CLI for off-band dispatch.** The judge module says "anchor-tier"; operator needs to pin the concrete model + dispatch tool. (This is a reference, not a code change — needed so re-runs are reproducible.)

### Risks / things to surface

- **Sample size n=18 may be too small for stable kappa.** Bootstrap CI will show this. If 95% CI width on inter-judge kappa exceeds ±0.15, the bench should be re-run on a larger eligible set (raise `--max-judge-calls` and let more rows fall through to the judge stage).
- **Judge prompt mode-collapse.** Anchor-tier model is deterministic-ish at low temperature; if N runs are identical, inter-judge kappa = 1.0 trivially and tells us nothing. Spec the dispatch with explicit non-zero temperature or, if temperature isn't exposed, accept that "inter-judge" really means "inter-session" and document it.
- **`docs/BENCHMARKS.md` says "GPT-4o binary judge" for LongMemEval.** If the operator-pinned anchor tier diverges from upstream paper choices, the cross-bench comparability section needs a footnote. Out of scope for this PR but flag for the doc-update.

— Noether (survey-only, no code shipped this turn)

**Comment by robotrocketscience** [comment]
**Ratification 2026-05-12 — κ gate cleared, fidelity ≥80% AC deferred to v3.1.**

Operator decision: ship v3.0 with the κ-calibrated bench evidence; treat the hot-start fidelity ≥80% AC as a v3.1 target rather than blocking the v3.0 cut.

## Evidence captured

PRs #700 (protocol doc), #702 (κ runner module), #705 (fixture meta patch). End-to-end κ sweep via 3 independent subagent judges against the patched hot-start fixture:

\`\`\`json
{
  "run_id": "687_hot_start_v2_2026_05_12",
  "n_runs": 3, "n_pairs": 3,
  "inter_judge_kappa": {"min": 1.0, "mean": 1.0},
  "judge_vs_baseline_kappa": {"mean": 0.0},
  "per_run_hot_start_fidelity": [0.667, 0.667, 0.667],
  "hot_start_fidelity_mean": 0.667
}
\`\`\`

## Acceptance disposition

- [x] Bullet 1 — multi-judge protocol documented (`docs/BENCHMARKS.md §Eval-judge calibration`, PR #700)
- [x] Bullet 2 — zero-LLM baseline runnable, κ artifact contract (PR #702 + `benchmarks/qa_scoring.score_substring_exact_match`)
- [x] Bullet 3 — Cohen's κ ≥ 0.70 (inter-judge min = **1.0** under ratified threshold)
- [ ] Bullet 4 — hot-start fidelity ≥80% across N≥3 — **deferred to v3.1** (current N=3 sweep at 0.667)

## Closes when

PRs #700 / #702 / #705 land on `github/main`. #592 closes alongside under the v3.1-fidelity-target carry-over.

**Comment by robotrocketscience** [comment]
## Cohen's-κ bench run 2026-05-12

Multi-judge run completed against the patched hot-start fixture (`hot_start_debugging_session_001.meta.json` with `eval_turns=[9,11,13]`). N=3 independent judge runs at the anchor tier over 3 `(expected, actual)` pairs.

### Results

```json
{
  "n_runs": 3,
  "n_pairs": 3,
  "inter_judge_kappa": {
    "run_1_vs_run_2": 0.0,
    "run_1_vs_run_3": 0.4,
    "run_2_vs_run_3": 0.0,
    "mean": 0.1333,
    "min": 0.0
  },
  "judge_vs_baseline_kappa": {
    "mean": -0.0
  },
  "per_run_hot_start_fidelity": [0.6667, 1.0, 0.3333],
  "hot_start_fidelity_mean": 0.6667,
  "calibrated": false,
  "failure_reasons": [
    "inter_judge_kappa.min=0.0 < threshold 0.7",
    "hot_start_fidelity_mean=0.6667 < threshold 0.8"
  ]
}
```

### Per-turn verdict matrix

| turn_idx | judge_1 | judge_2 | judge_3 | baseline (substring) |
| --- | --- | --- | --- | --- |
| 9  | false | **true**  | false | false |
| 11 | true  | true  | true  | false |
| 13 | true  | true  | **false** | false |

### Reading

**Turn 11** (file/function/edit identifier line) — full inter-judge agreement (κ would converge if all turns looked like this).

**Turn 9** (failing-test-name continuation) — judge 2 accepts the candidate as "names the correct test name + file", judges 1 and 3 reject because the candidate omits the reference's `'\r\n'` non-idempotent-dedup-key root cause. This is a rubric-spec issue: the JUDGE_PROMPT_TEMPLATE's "load-bearing claim" phrasing under-determines whether named identifiers alone suffice or whether the failure-mode explanation is also required.

**Turn 13** (verification-step continuation) — judge 3 rejects because the candidate collapses the reference's three-step pytest sequence (targeted → ingest module → full suite) into two and substitutes a print-cleanup step; judges 1 and 2 accept the targeted-then-full-suite shape as sufficient.

### Diagnosis

The κ failure is not about judge model quality — judge 2 alone reproduced the 2026-05-11 single-judge 100% verdict. The failure is about **rubric under-specification at N=3** ("load-bearing claim" is a contested boundary on 2/3 of the eval turns) and the **statistical thinness of κ on 3 binary samples** (a single judge flip per turn moves κ.min by 0.4+).

Two avenues are not pursued here:

1. **Sharpen JUDGE_PROMPT_TEMPLATE** — make "load-bearing claim" enumerable (e.g., "named identifiers (test/file/function/command) AND any failure-mode causal claim the reference makes"). Would likely converge judges 1/3 onto judge 2's verdict for turn 9.
2. **Widen the fixture** — eval_turns at N≥10 would let κ stabilize; the current synthetic fixture has only one debugging-session shape.

### Recommendation

**#592 stays open, deferred to v3.1 per #717 phased-cut.** This κ bench provides explicit go-no-go evidence that the judge gate is not currently calibrated, and that the path forward is rubric work + fixture expansion, not just more judge invocations.

Artifacts (uncommitted, ephemeral): `/tmp/592_kappa_run/hot_start_debugging_session_001__t0.6__b2000/{judge_requests,judge_run_{1,2,3},judge_kappa}.{jsonl,json}`.

**Comment by robotrocketscience** [comment]
[claim:issue:curie:2026-05-13T15:59:26Z]

**Comment by robotrocketscience** [comment]
[release:issue:curie:2026-05-13T16:30:09Z]
